src/test/merkle_tests.cpp

```diff
@@ -24,9 +24,48 @@ extern Array read_json(const std::string& jsondata);
 
 using namespace std;
 
+template<typename T>
+void expect_deser_same(const T& expected)
+{
+    CDataStream ss1(SER_NETWORK, PROTOCOL_VERSION);
+    ss1 << expected;
+
+    auto serialized_size = ss1.size();
+
+    T object;
+    ss1 >> object;
+
+    CDataStream ss2(SER_NETWORK, PROTOCOL_VERSION);
+    ss2 << object;
+
+    BOOST_CHECK(serialized_size == ss2.size());
+    BOOST_CHECK(memcmp(&*ss1.begin(), &*ss2.begin(), serialized_size) == 0);
+}
+
+template<>
+void expect_deser_same(const ZCTestingIncrementalWitness& expected)
+{
+    // Cannot check this; IncrementalWitness cannot be
+    // deserialized because it can only be constructed by
+    // IncrementalMerkleTree, and it does not yet have a
+    // canonical serialized representation.
+}
+
+template<>
+void expect_deser_same(const libzcash::MerklePath& expected)
+{
+    // This deserialization check is pointless for MerklePath,
+    // since we only serialize it to check it against test
+    // vectors. See `expect_test_vector` for that. Also,
+    // it doesn't seem that vector<bool> can be properly
+    // deserialized by Bitcoin's serialization code.
```

Comment:
I thought of that too, but unfortunately it simply cannot be compiled. It's not a bug, it's a problem with `vector<bool>`'s specialization in STL which is not handled by Bitcoin's serialization code somehow. At least that's what I think, but I couldn't read the mess of template substitution errors.